___

<center><img src="../HAP-E-logo.png" alt="HAP-E" width="200"/></center>
  
___
<em>HAP-E Group tutorial</em>

---
# Print formatting with strings and variable
---

In this tutorial, we will learn about how to format output with strings and other data types. 

A crude way of constructing an output string would be to add together all of the pieces of information you want to display in a sequence, or *concatentation*.  This solves your problem once, but it is better to solve problems like that programmatically.  That is where string formatting comes in.

As a quick comparison, consider:



`rider = 'Rossi'
gp_victories = 115`

`'In his career, '+rider+' won '+str(gp_victories)+' Grand Prix victories.'  # concatenation`

`f'In his career, {player} won {points} Grand Prix victories..'    # string formatting`


There are several ways to perform string formatting which you are likely to encounter in code, and we will practice a few examples here.

We'll cover the following topics:

    1. Formatting with placeholders (using the modulo `%` character, a bit old fashioned) 
    2. The `.format()` method (a workhorse, programmatic method)
    3. The `f-strings` method (newest and probably preferred when writing new code)



---
## 1. Formatting with placeholders
---

You can use <code>%my_string</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".  There are several kinds of formatting that we can use that might be useful in different situations and we will look at a few of them here.

In [5]:
# Try this:
format_statement = "My favourite %s is %s"
mammal_info = ("mammal", "the Dikdik")
bird_info = ("bird", "the Eurasian Jay")

print(format_statement % mammal_info)

My favourite mammal is the Dikdik


You can pass multiple items manually by placing them inside a tuple after the `%` operator.

In [7]:
# "Manual" way
print("I'm going to %s first, and then I'm going to %s" %('Kenya','Tanzania'))

I'm going to Kenya first, and then I'm going to Tanzania


You can also pass variable names:

In [8]:
# "Coded" way
x, y = 'Kenya','Tanzania'
print("I'm going to %s first, and then I'm going to %s"%(x,y))

I'm going to Kenya first, and then I'm going to Tanzania


### Format conversion methods.
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`, respectively. We may learn more about these functions in the future, but for now it is enough to note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [11]:
# Try this:
print('He said to call him %s.' %'Ishmail')
print('He said to call him %r.' %'Ishmail')

He said to call him Ishmail.
He said to call him 'Ishmail'.


As another example, `\t` inserts a tab into a string.

In [14]:
# %s and %r handle this differently
print('The white whale was %s.' %'this \tbig')
print('The white whale was %r.' %'this \tbig')

The white whale was this 	big.
The white whale was 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [16]:
print('I watched %s nature shows today.' %3.75)
print('I watched %d nature shows today.' %3.75)   

I watched 3.75 nature shows today.
I watched 3 nature shows today.


### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain and these may be padded with whitespace if the entire number does not have the indicated number of digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [17]:
print('Floating point numbers gone wild: %5.2f' %(13.144))

Floating point numbers gone wild: 13.14


In [18]:
print('Floating point numbers gone less wild: %1.0f' %(13.144))

Floating point numbers gone less wild: 13


In [9]:
print('Floating point numbers gone wilder: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [10]:
print('Floating point numbers gone long: %10.2f' %(13.144))

Floating point numbers:      13.14


In [11]:
print('Floating point numbers gone longer: %25.2f' %(13.144))

Floating point numbers:                     13.14


### Multiple Formatting
Finally, nothing stops us from using more than one conversion tool in the same print statement:

In [20]:
print('I said %s, not %5.2f. %r' %('Pie',3.1415,'Bye!'))

I said Pie, not  3.14. 'Bye!'


For more information on string formatting with "old fashioned" placeholders, you can look at the official Python documentation https://docs.python.org/3/library/stdtypes.html#old-string-formatting

## 2. The `.format()` method
A modern, "programmatic" way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

`Insert {} then also add {}'.format('something1','something2')`
    
For example:

In [23]:
# I hope you like memes
print('This is a string with an {}'.format('...My name\'s Jeeeeeff!'))

This is a string with an ...My name's Jeeeeeff!


### The .format() method has several advantages over the %s placeholder method:

#### A. Inserted objects can be called by index position:

In [24]:
print('Panda {2} {1} and {0}'.format('leaves','shoots','eats'))

Panda eats shoots and leaves


#### B. Inserted objects can be assigned to and called by keywords:

In [25]:
print('First value: {a}, Second value: {b}, Third value: {c}'.format(a=1, b='Two', c=12.3))

First value: 1, Second value: Two, Third value: 12.3


#### C. Inserted objects can be reused, avoiding duplication:

In [27]:
# one solution
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
# improved solution with .format()
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with `.format()`
Within the curly braces `{}` you can assign field lengths, left and right alignment, rounding parameters and more

In [34]:
print('{0:8} | {1:9}'.format('Fruit', 'Amount'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Amount   
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [18]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can also choose to precede the aligment operator with a padding character (e.g. like you may for a formatted index)

In [19]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [20]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

A lot more detail is possible but this is a good introduction. For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

## 3. The `f`-strings method

A more modern method for string formatting, `f`-strings offer several benefits over the `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [35]:
name = 'Ishmail'

print(f"He said his name is {name}.")

He said his name is Ishmail.


You can use the argument `!r` to get the string representation version:

In [36]:
print(f"He said his name is {name!r}")

He said his name is 'Ishmail'


#### Float formatting follows `"result: {value:{width}.{precision}}"`

With the `.format()` method you might see `{value:10.4f}`, with f-strings this may become `{value:{10}.{6}}`


In [39]:
num = 23.45678

# .format() - 10 spaces, 4 decimals
print("My 10 character, four decimal number is:{0:10.4f}".format(num))

# f-string version 10 spaces, 6 total digits
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


NB with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. 

However, f-strings do not pad to the right of the decimal, even if precision allows it:

In [43]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [44]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

That is the basics of string formatting!